File: app/src/main/java/com/kobrakid/retroachievements/adapter/GameCommentsAdapter.java
package com.kobrakid.retroachievements.adapter;

import android.content.Context;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.kobrakid.retroachievements.Consts;
import com.kobrakid.retroachievements.R;
import com.squareup.picasso.Picasso;

import java.util.List;
import java.util.Map;
import java.util.Objects;

public class GameCommentsAdapter extends RecyclerView.Adapter {

    private Context context;
    private Map<String, List<String>> comments;

    public GameCommentsAdapter(Context context, Map<String, List<String>> comments) {
        this.context = context;
        this.comments = comments;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new GameCommentViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.view_holder_game_comment, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        Picasso.get()
                .load(Consts.BASE_URL
                        + "/" + Consts.USER_PIC_POSTFIX
                        + "/" + Objects.requireNonNull(comments.get("user")).get(position) + ".png")
                .placeholder(R.drawable.user_placeholder)
                .into(((ImageView) holder.itemView.findViewById(R.id.comments_user_icon)));
        ((TextView) holder.itemView.findViewById(R.id.comments_user_comment))
                .setText(Objects.requireNonNull(comments.get("text")).get(position));
        ((TextView) holder.itemView.findViewById(R.id.comments_user_name))
                .setText(Objects.requireNonNull(comments.get("user")).get(position));
        ((TextView) holder.itemView.findViewById(R.id.comments_date))
                .setText(Objects.requireNonNull(comments.get("date")).get(position));
        ((TextView) holder.itemView.findViewById(R.id.comments_user_rank_score))
                .setText(context.getString(R.string.score_rank, Objects.requireNonNull(comments.get("score")).get(position), Objects.requireNonNull(comments.get("rank")).get(position)));
        if (!Objects.requireNonNull(comments.get("tag")).get(position).equals("RA_NO_TAG_" + position)) {
            ((TextView) holder.itemView.findViewById(R.id.comments_user_tag))
                    .setText(context.getString(R.string.quote, Objects.requireNonNull(comments.get("tag")).get(position)));
            holder.itemView.findViewById(R.id.comments_user_tag).setVisibility(View.VISIBLE);
        } else {
            holder.itemView.findViewById(R.id.comments_user_tag).setVisibility(View.GONE);
        }
        TextView username = holder.itemView.findViewById(R.id.comments_user_name);
        switch (comments.get("acct").get(position)) {
            case "Banned":
                username.setTextColor(Color.RED);
                username.setPaintFlags(username.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
                break;
            case "Developer":
                username.setTextColor(Color.GREEN);
                username.setPaintFlags(username.getPaintFlags() & (~Paint.STRIKE_THRU_TEXT_FLAG));
                break;
            case "Registered":
                final TypedValue primaryColor = new TypedValue();
                context.getTheme().resolveAttribute(R.attr.colorPrimary, primaryColor, true);
                username.setTextColor(primaryColor.data);
                username.setPaintFlags(username.getPaintFlags() & (~Paint.STRIKE_THRU_TEXT_FLAG));
                break;
            default:
                username.setTextColor(Color.YELLOW);
                username.setPaintFlags(username.getPaintFlags() & (~Paint.STRIKE_THRU_TEXT_FLAG));
        }
    }

    @Override
    public int getItemCount() {
        return comments.get("text") != null ? Objects.requireNonNull(comments.get("text")).size() : 0;
    }

    static class GameCommentViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        GameCommentViewHolder(@NonNull View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            View extra = view.findViewById(R.id.comment_hide_content);
            if (extra != null) {
                extra.setVisibility(extra.getVisibility() == View.VISIBLE ? View.GONE : View.VISIBLE);
                ((TextView) view.findViewById(R.id.comments_user_comment)).setMaxLines(extra.getVisibility() == View.VISIBLE ? 10 : 3);
            }
        }
    }

}

File: app/src/main/java/com/kobrakid/retroachievements/adapter/GameCommentsAdapter.kt
package com.kobrakid.retroachievements.adapter

import android.graphics.Color
import android.graphics.Paint
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.kobrakid.retroachievements.Consts
import com.kobrakid.retroachievements.R
import com.squareup.picasso.Picasso
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.withContext

class GameCommentsAdapter : RecyclerView.Adapter<GameCommentsAdapter.GameCommentViewHolder>() {

    private val comments = mutableListOf<String>()

    // Users' profile pictures
    private val users = mutableListOf<String>()

    // Users' account statuses
    private val accounts = mutableListOf<String>()
    private val scores = mutableListOf<String>()
    private val ranks = mutableListOf<String>()

    // Users' taglines
    private val tags = mutableListOf<String>()
    private val dates = mutableListOf<String>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GameCommentViewHolder {
        return GameCommentViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.view_holder_game_comment, parent, false))
    }

    override fun onBindViewHolder(holder: GameCommentViewHolder, position: Int) {
        Picasso.get()
                .load(Consts.BASE_URL + "/" + Consts.USER_PIC_POSTFIX + "/" + users[position] + ".png")
                .placeholder(R.drawable.user_placeholder)
                .into(holder.itemView.findViewById<ImageView>(R.id.comments_user_icon))
        holder.itemView.findViewById<TextView>(R.id.comments_user_comment).text = comments[position]
        holder.itemView.findViewById<TextView>(R.id.comments_date).text = dates[position]
        holder.itemView.findViewById<TextView>(R.id.comments_user_rank_score).text = holder.itemView.context.getString(R.string.score_rank, scores[position], ranks[position])
        with(holder.itemView.findViewById<TextView>(R.id.comments_user_tag)) {
            when (tags[position]) {
                "_RA_NO_TAG" -> visibility = View.GONE
                else -> {
                    text = holder.itemView.context.getString(R.string.quote, tags[position])
                    visibility = View.VISIBLE
                }
            }
        }
        with(holder.itemView.findViewById<TextView>(R.id.comments_user_name)) {
            text = users[position]
            when (accounts[position]) {
                "Banned" -> {
                    setTextColor(Color.RED)
                    paintFlags = paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
                }
                "Developer" -> {
                    setTextColor(Color.GREEN)
                    paintFlags = paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
                }
                "Registered" -> {
                    val primaryColor = TypedValue()
                    holder.itemView.context.theme.resolveAttribute(R.attr.colorPrimary, primaryColor, true)
                    setTextColor(primaryColor.data)
                    paintFlags = paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
                }
                else -> {
                    setTextColor(Color.YELLOW)
                    paintFlags = paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return comments.size
    }

    suspend fun addComment(text: String, user: String, account: String, score: String, rank: String, tag: String, date: String) {
        withContext(Main) {
            comments.add(text)
            users.add(user)
            accounts.add(account)
            scores.add(score)
            ranks.add(rank)
            tags.add(tag)
            dates.add(date)
            notifyItemInserted(comments.size - 1)
        }
    }

    class GameCommentViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        override fun onClick(view: View) {
            val extra = view.findViewById<View>(R.id.comment_hide_content)
            if (extra != null) {
                extra.visibility = if (extra.visibility == View.VISIBLE) View.GONE else View.VISIBLE
                view.findViewById<TextView>(R.id.comments_user_comment).maxLines = if (extra.visibility == View.VISIBLE) 10 else 3
            }
        }

        init {
            itemView.setOnClickListener(this)
        }
    }

}